src/main.rs

```diff
@@ -1,3 +1,196 @@
+}
+
+#[repr(u8)]
+#[derive(Debug)]
+enum RespType {
+    RespOk = 0,
+    RespErr,
+    NotifyOutputAvailable,
+    NotifyProcessDied,
+}
+
+/* Is there a better way? */
+impl RespType {
+    fn from_u8(x: u8) -> Option<RespType> {
+        match x {
+            0 => Some(RespType::RespOk),
+            1 => Some(RespType::RespErr),
+            2 => Some(RespType::NotifyOutputAvailable),
+            3 => Some(RespType::NotifyProcessDied),
+            _ => None,
+        }
+    }
+}
+
+struct GuestAgent {
+    stream: UnixStream,
+    last_msg_id: u64,
+}
+
+impl GuestAgent {
+    fn connected(path: &str) -> Result<GuestAgent, io::Error> {
+        loop {
+            match UnixStream::connect(path) {
+                Ok(s) => {
+                    break Ok(GuestAgent {
+                        stream: s,
+                        last_msg_id: 0,
+                    })
+                }
+                Err(err) => match err.kind() {
+                    io::ErrorKind::NotFound => {
+                        println!("Waiting for Guest Agent socket ...");
+                        thread::sleep(time::Duration::from_secs(1));
+                    }
+                    _ => break Err(err),
+                },
+            };
+        }
+    }
+
+    fn get_new_msg_id(&mut self) -> u64 {
+        self.last_msg_id += 1;
+        self.last_msg_id
+    }
+
+    fn create_header(msg_id: u64, typ: MsgType) -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend_from_slice(&msg_id.to_le_bytes());
+        v.extend_from_slice(&(typ as u8).to_le_bytes());
+        v
+    }
+
+    fn append_submsg_type(buf: &mut Vec<u8>, subtype: MsgSubType) -> () {
+        buf.extend(&(subtype as u8).to_le_bytes());
+    }
+
+    fn append_submsg_u32(buf: &mut Vec<u8>, subtype: MsgSubType, val: u32) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        buf.extend_from_slice(&val.to_le_bytes());
+    }
+
+    fn append_bytes(buf: &mut Vec<u8>, bytes: &[u8]) -> () {
+        buf.extend_from_slice(&bytes.len().to_le_bytes());
+        buf.extend_from_slice(bytes);
+    }
+
+    fn append_submsg_bytes(buf: &mut Vec<u8>, subtype: MsgSubType, bytes: &[u8]) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        GuestAgent::append_bytes(buf, bytes);
+    }
+
+    fn append_submsg_array(buf: &mut Vec<u8>, subtype: MsgSubType, array: &[&[u8]]) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        buf.extend_from_slice(&array.len().to_le_bytes());
+        for &a in array {
+            GuestAgent::append_bytes(buf, &a);
+        }
+    }
+
+    fn get_response(&mut self) -> Result<(), io::Error> {
+        let mut buf = [0; 8];
+        self.stream.read_exact(&mut buf)?;
+        let id = u64::from_le_bytes(buf);
+        println!("Message ID: {}", id);
+
+        let mut buf = [0; 1];
+        self.stream.read_exact(&mut buf)?;
+        let typ = RespType::from_u8(buf[0]).expect("Invalid response type!");
```

Comment:
You usually want to propagate errors (like the `io::Error`). With a custom error type you could

```suggestion
        let typ = RespType::try_from(buf[0])?;
```

and handle it / panic in `main`.

Reply:
Done.